core/ir/category_api.h

```diff
@@ -0,0 +1,15 @@
+#ifndef _DR_IR_CATEGORIES_H_
+#define _DR_IR_CATEGORIES_H_ 1
+
+typedef enum _dr_instr_category_type_t {
+    DR_INSTR_CATEGORY_UNCATEGORIZED = 0x0, /* Uncategorised. */
+    DR_INSTR_CATEGORY_MATH_INT = 0x1,      /* Integer arithmetic operations */
+    DR_INSTR_CATEGORY_MATH_FLOAT = 0x2,    /* Floating-Point arithmetic operations */
+    DR_INSTR_CATEGORY_LOAD = 0x4,          /* Loads */
+    DR_INSTR_CATEGORY_STORE = 0x8,         /* Stores */
+    DR_INSTR_CATEGORY_BRANCH = 0x10,       /* Branches */
+    DR_INSTR_CATEGORY_SIMD = 0x20,         /* Operations with vector registers (SIMD) */
+    DR_INSTR_CATEGORY_OTHER = 0x40         /* Other types of instructions */
```

Comment:
What about instr_is_floating_ex() and dr_fp_type t (which seems to be completely missing for AArch64 today)?  I assume that would be implemented using the same category system: so we would want these types to also split into those categories?

Reply:
We will add on the next step. Need to discuss and fix that patch )))
Also we will share x86 approach. Uunfortunantly it is not programming just something like
```
[core/ir/x86/decode_table.c](https://github.com/DynamoRIO/dynamorio/commit/40cd2fa7cc53e54667b34430e8d5274bf713cecd#diff-f9b347cfa07cc46b1ea538b6e01652a0bdac4ad6c868183d734e4d823fa76853)
@@ -10703,4 +10703,1440 @@ const instr_info_t extra_operands[] =
    {OP_CONTD, 0xc00f0171, "<enclv cont'd>", ecx, edx, edx, xx, xx, mrm, x, END_LIST},
};

const uint category_by_opcode[] = {
    /* OP_INVALID */           DR_INSTR_CATEGORY_OTHER,
    /* OP_UNDECODED */         DR_INSTR_CATEGORY_OTHER,
    /* OP_CONTD */             DR_INSTR_CATEGORY_OTHER,
    /* OP_LABEL */             DR_INSTR_CATEGORY_OTHER,
    /* OP_add */               DR_INSTR_CATEGORY_MATH_INT,
   ...
```